File: backend/lib/yarnballs/explosion.ex
defmodule Yarnballs.Explosions do
  @moduledoc """
  Represents a collection of explosions.
  """
  alias Yarnballs.Explosion
  alias Yarnballs.Enemy

  @type t :: %__MODULE__{
          entities: list(Explosion.t())
        }

  @enforce_keys [:entities]
  @derive {Jason.Encoder, only: [:entities]}
  defstruct @enforce_keys

  def init() do
    %__MODULE__{entities: []}
  end

  def spawn(explosions, thing) do
    %{explosions | entities: [Enemy.explode(thing) | explosions.entities]}
  end

  def update(explosions) do
    entities =
      explosions.entities
      |> Enum.filter(&keep?/1)
      |> Enum.map(&Explosion.update/1)

    %{explosions | entities: entities}
  end

  defp keep?(explosion) do
    explosion.lifespan > 0
  end
end

defmodule Yarnballs.Explosion do
  @moduledoc """
  Represent an explosion.
  """
  require Logger

  @type t :: %__MODULE__{
          id: binary,
          updated_at: integer,
          x: float,
          y: float,
          size: float,
          lifespan: integer
        }

  @enforce_keys [
    :id,
    :updated_at,
    :x,
    :y,
    # the size of the thing that is exploding - used to determine scale.
    :size,
    :lifespan
  ]
  @derive {Jason.Encoder, only: @enforce_keys}
  defstruct @enforce_keys

  @lifespan 1000

  def spawn(x, y, size) do
    %__MODULE__{
      id: Ecto.UUID.generate(),
      updated_at: Yarnballs.Utils.now_milliseconds(),
      x: x,
      y: y,
      size: size,
      lifespan: @lifespan
    }
  end

  def update(explosion) do
    # update physics
    updated_at = Yarnballs.Utils.now_milliseconds()
    dt = updated_at - explosion.updated_at
    lifespan = explosion.lifespan - dt

    %{explosion | updated_at: updated_at, lifespan: lifespan}
  end
end

File: yarnballs/native/yarnballs/src/explosion.rs
use crate::enemy::Enemy;
use crate::utils;
use rustler::NifStruct;

#[derive(NifStruct)]
#[module = "Yarnballs.Explosions"]
pub struct Explosions {
    entities: Vec<Explosion>,
}

impl Explosions {
    pub fn init() -> Self {
        Self {
            entities: Vec::new(),
        }
    }

    pub fn spawn(&mut self, enemy: &Enemy) {
        let explosion = enemy.explode();
        self.entities.push(explosion)
    }

    pub fn update(&mut self) {
        self.entities.retain(|e| e.lifespan > 0);
        self.entities.iter_mut().for_each(|e| e.update());
    }
}

#[derive(NifStruct)]
#[module = "Yarnballs.Explosion"]
pub struct Explosion {
    id: String,
    updated_at: i64,
    x: f64,
    y: f64,
    // the size of the thing that is exploding - used to determine scale.
    size: f64,
    lifespan: i64,
}

const LIFESPAN: i64 = 1000;

impl Explosion {
    pub fn spawn(x: f64, y: f64, size: f64) -> Self {
        Self {
            id: utils::new_uuid(),
            updated_at: utils::now_in_millis(),
            x,
            y,
            size,
            lifespan: LIFESPAN,
        }
    }

    pub fn update(&mut self) {
        let updated_at = utils::now_in_millis();
        let dt = updated_at - self.updated_at;

        self.lifespan -= dt;
        self.updated_at = updated_at;
    }
}
